// istanbul-lib-instrument/dist/dist/read-coverage.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = readInitialCoverage;

var _core = require("@babel/core");

var _schema = require("@istanbuljs/schema");

var _constants = require("./constants");

function getAst(code) {
  if (_typeof(code) === 'object' && typeof code.type === 'string') {
    // Assume code is already a babel ast.
    return code;
  }

  if (typeof code !== 'string') {
    throw new Error('Code must be a string');
  } // Parse as leniently as possible


  return (0, _core.parseSync)(code, {
    babelrc: false,
    configFile: false,
    parserOpts: {
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'script',
      plugins: _schema.defaults.instrumenter.parserPlugins
    }
  });
}

function readInitialCoverage(code) {
  var ast = getAst(code);
  var covScope;
  (0, _core.traverse)(ast, {
    ObjectProperty: function ObjectProperty(path) {
      var node = path.node;

      if (!node.computed && path.get('key').isIdentifier() && node.key.name === _constants.MAGIC_KEY) {
        var magicValue = path.get('value').evaluate();

        if (!magicValue.confident || magicValue.value !== _constants.MAGIC_VALUE) {
          return;
        }

        covScope = path.scope.getFunctionParent() || path.scope.getProgramParent();
        path.stop();
      }
    }
  });

  if (!covScope) {
    return null;
  }

  var result = {};

  for (var _i = 0, _arr = ['path', 'hash', 'gcv', 'coverageData']; _i < _arr.length; _i++) {
    var key = _arr[_i];
    var binding = covScope.getOwnBinding(key);

    if (!binding) {
      return null;
    }

    var valuePath = binding.path.get('init');
    var value = valuePath.evaluate();

    if (!value.confident) {
      return null;
    }

    result[key] = value.value;
  }

  delete result.coverageData[_constants.MAGIC_KEY];
  delete result.coverageData.hash;
  return result;
}
