// jsdom/lib/jsdom/living/nodes/dist/HTMLTemplateElement-impl.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var HTMLElementImpl = require("./HTMLElement-impl").implementation;

var Document = require("../generated/Document");

var DocumentFragment = require("../generated/DocumentFragment");

var _require = require("../helpers/internal-constants"),
    cloningSteps = _require.cloningSteps,
    domSymbolTree = _require.domSymbolTree;

var _require2 = require("../node"),
    clone = _require2.clone;

var HTMLTemplateElementImpl =
/*#__PURE__*/
function (_HTMLElementImpl) {
  _inherits(HTMLTemplateElementImpl, _HTMLElementImpl);

  function HTMLTemplateElementImpl(globalObject, args, privateData) {
    var _this;

    _classCallCheck(this, HTMLTemplateElementImpl);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(HTMLTemplateElementImpl).call(this, globalObject, args, privateData));

    var doc = _this._appropriateTemplateContentsOwnerDocument(_this._ownerDocument);

    _this._templateContents = DocumentFragment.createImpl(_this._globalObject, [], {
      ownerDocument: doc,
      host: _assertThisInitialized(_this)
    });
    return _this;
  } // https://html.spec.whatwg.org/multipage/scripting.html#appropriate-template-contents-owner-document


  _createClass(HTMLTemplateElementImpl, [{
    key: "_appropriateTemplateContentsOwnerDocument",
    value: function _appropriateTemplateContentsOwnerDocument(doc) {
      if (!doc._isInertTemplateDocument) {
        if (doc._associatedInertTemplateDocument === undefined) {
          var newDoc = Document.createImpl(this._globalObject, [], {
            options: {
              parsingMode: doc._parsingMode,
              encoding: doc._encoding
            }
          });
          newDoc._isInertTemplateDocument = true;
          doc._associatedInertTemplateDocument = newDoc;
        }

        doc = doc._associatedInertTemplateDocument;
      }

      return doc;
    } // https://html.spec.whatwg.org/multipage/scripting.html#template-adopting-steps

  }, {
    key: "_adoptingSteps",
    value: function _adoptingSteps() {
      var doc = this._appropriateTemplateContentsOwnerDocument(this._ownerDocument);

      doc._adoptNode(this._templateContents);
    }
  }, {
    key: cloningSteps,
    value: function value(copy, node, document, cloneChildren) {
      if (!cloneChildren) {
        return;
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = domSymbolTree.childrenIterator(node._templateContents)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var child = _step.value;
          var childCopy = clone(child, copy._templateContents._ownerDocument, true);

          copy._templateContents.appendChild(childCopy);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "content",
    get: function get() {
      return this._templateContents;
    }
  }]);

  return HTMLTemplateElementImpl;
}(HTMLElementImpl);

module.exports = {
  implementation: HTMLTemplateElementImpl
};
