// whatwg-url/dist/dist/URL.dev.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;
var interfaceName = "URL";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'URL'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["URL"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor URL is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window", "Worker"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  var URL =
  /*#__PURE__*/
  function () {
    function URL(url) {
      _classCallCheck(this, URL);

      if (arguments.length < 1) {
        throw new TypeError("Failed to construct 'URL': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["USVString"](curArg, {
          context: "Failed to construct 'URL': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];

        if (_curArg !== undefined) {
          _curArg = conversions["USVString"](_curArg, {
            context: "Failed to construct 'URL': parameter 2"
          });
        }

        args.push(_curArg);
      }
      return exports.setup(Object.create((this instanceof URL ? this.constructor : void 0).prototype), globalObject, args);
    }

    _createClass(URL, [{
      key: "toJSON",
      value: function toJSON() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'toJSON' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol].toJSON();
      }
    }, {
      key: "toString",
      value: function toString() {
        var esValue = this;

        if (!exports.is(esValue)) {
          throw new TypeError("'toString' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["href"];
      }
    }, {
      key: "href",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get href' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["href"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set href' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'href' property on 'URL': The provided value"
        });
        esValue[implSymbol]["href"] = V;
      }
    }, {
      key: "origin",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get origin' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["origin"];
      }
    }, {
      key: "protocol",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get protocol' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["protocol"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set protocol' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'protocol' property on 'URL': The provided value"
        });
        esValue[implSymbol]["protocol"] = V;
      }
    }, {
      key: "username",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get username' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["username"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set username' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'username' property on 'URL': The provided value"
        });
        esValue[implSymbol]["username"] = V;
      }
    }, {
      key: "password",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get password' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["password"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set password' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'password' property on 'URL': The provided value"
        });
        esValue[implSymbol]["password"] = V;
      }
    }, {
      key: "host",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get host' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["host"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set host' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'host' property on 'URL': The provided value"
        });
        esValue[implSymbol]["host"] = V;
      }
    }, {
      key: "hostname",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get hostname' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["hostname"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set hostname' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'hostname' property on 'URL': The provided value"
        });
        esValue[implSymbol]["hostname"] = V;
      }
    }, {
      key: "port",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get port' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["port"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set port' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'port' property on 'URL': The provided value"
        });
        esValue[implSymbol]["port"] = V;
      }
    }, {
      key: "pathname",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get pathname' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["pathname"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set pathname' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'pathname' property on 'URL': The provided value"
        });
        esValue[implSymbol]["pathname"] = V;
      }
    }, {
      key: "search",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get search' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["search"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set search' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'search' property on 'URL': The provided value"
        });
        esValue[implSymbol]["search"] = V;
      }
    }, {
      key: "searchParams",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get searchParams' called on an object that is not a valid instance of URL.");
        }

        return utils.getSameObject(this, "searchParams", function () {
          return utils.tryWrapperForImpl(esValue[implSymbol]["searchParams"]);
        });
      }
    }, {
      key: "hash",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'get hash' called on an object that is not a valid instance of URL.");
        }

        return esValue[implSymbol]["hash"];
      },
      set: function set(V) {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("'set hash' called on an object that is not a valid instance of URL.");
        }

        V = conversions["USVString"](V, {
          context: "Failed to set the 'hash' property on 'URL': The provided value"
        });
        esValue[implSymbol]["hash"] = V;
      }
    }]);

    return URL;
  }();

  Object.defineProperties(URL.prototype, _defineProperty({
    toJSON: {
      enumerable: true
    },
    href: {
      enumerable: true
    },
    toString: {
      enumerable: true
    },
    origin: {
      enumerable: true
    },
    protocol: {
      enumerable: true
    },
    username: {
      enumerable: true
    },
    password: {
      enumerable: true
    },
    host: {
      enumerable: true
    },
    hostname: {
      enumerable: true
    },
    port: {
      enumerable: true
    },
    pathname: {
      enumerable: true
    },
    search: {
      enumerable: true
    },
    searchParams: {
      enumerable: true
    },
    hash: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "URL",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = URL;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: URL
  });

  if (globalNames.includes("Window")) {
    Object.defineProperty(globalObject, "webkitURL", {
      configurable: true,
      writable: true,
      value: URL
    });
  }
};

var Impl = require("./URL-impl.js");
