// jest-jasmine2/build/jasmine/dist/jasmineLight.dev.js
'use strict';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports._interface = exports.create = void 0;

var _Env = _interopRequireDefault(require('./Env'));

var _JsApiReporter = _interopRequireDefault(require('./JsApiReporter'));

var _ReportDispatcher = _interopRequireDefault(require('./ReportDispatcher'));

var _Spec = _interopRequireDefault(require('./Spec'));

var _Suite = _interopRequireDefault(require('./Suite'));

var _Timer = _interopRequireDefault(require('./Timer'));

var _createSpy = _interopRequireDefault(require('./createSpy'));

var _spyRegistry = _interopRequireDefault(require('./spyRegistry'));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
// This file is a heavily modified fork of Jasmine. Original license:

/*
Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/* eslint-disable sort-keys, local/prefer-spread-eventually, local/prefer-rest-params-eventually */


var create = function create(createOptions) {
  var j$ = _objectSpread({}, createOptions);

  j$._DEFAULT_TIMEOUT_INTERVAL = createOptions.testTimeout || 5000;

  j$.getEnv = function (options) {
    var env = j$.currentEnv_ = j$.currentEnv_ || new j$.Env(options); //jasmine. singletons in here (setTimeout blah blah).

    return env;
  };

  j$.createSpy = _createSpy["default"];
  j$.Env = (0, _Env["default"])(j$);
  j$.JsApiReporter = _JsApiReporter["default"];
  j$.ReportDispatcher = _ReportDispatcher["default"];
  j$.Spec = _Spec["default"];
  j$.SpyRegistry = _spyRegistry["default"];
  j$.Suite = _Suite["default"];
  j$.Timer = _Timer["default"];
  j$.version = '2.5.2-light';
  return j$;
}; // Interface is a reserved word in strict mode, so can't export it as ESM


exports.create = create;

var _interface = function _interface(jasmine, env) {
  var jasmineInterface = {
    describe: function describe(description, specDefinitions) {
      return env.describe(description, specDefinitions);
    },
    xdescribe: function xdescribe(description, specDefinitions) {
      return env.xdescribe(description, specDefinitions);
    },
    fdescribe: function fdescribe(description, specDefinitions) {
      return env.fdescribe(description, specDefinitions);
    },
    it: function it() {
      return env.it.apply(env, arguments);
    },
    xit: function xit() {
      return env.xit.apply(env, arguments);
    },
    fit: function fit() {
      return env.fit.apply(env, arguments);
    },
    beforeEach: function beforeEach() {
      if (typeof arguments[0] !== 'function') {
        throw new Error('Invalid first argument. It must be a callback function.');
      }

      return env.beforeEach.apply(env, arguments);
    },
    afterEach: function afterEach() {
      if (typeof arguments[0] !== 'function') {
        throw new Error('Invalid first argument. It must be a callback function.');
      }

      return env.afterEach.apply(env, arguments);
    },
    beforeAll: function beforeAll() {
      if (typeof arguments[0] !== 'function') {
        throw new Error('Invalid first argument. It must be a callback function.');
      }

      return env.beforeAll.apply(env, arguments);
    },
    afterAll: function afterAll() {
      if (typeof arguments[0] !== 'function') {
        throw new Error('Invalid first argument. It must be a callback function.');
      }

      return env.afterAll.apply(env, arguments);
    },
    pending: function pending() {
      return env.pending.apply(env, arguments);
    },
    fail: function fail() {
      return env.fail.apply(env, arguments);
    },
    spyOn: function spyOn(obj, methodName, accessType) {
      return env.spyOn(obj, methodName, accessType);
    },
    jsApiReporter: new jasmine.JsApiReporter({
      timer: new jasmine.Timer()
    }),
    jasmine: jasmine
  };
  return jasmineInterface;
};

exports._interface = _interface;
