// @jest/console/build/dist/CustomConsole.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports["default"] = void 0;

function _assert() {
  var data = _interopRequireDefault(require('assert'));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _console() {
  var data = require('console');

  _console = function _console() {
    return data;
  };

  return data;
}

function _util() {
  var data = require('util');

  _util = function _util() {
    return data;
  };

  return data;
}

function _chalk() {
  var data = _interopRequireDefault(require('chalk'));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _jestUtil() {
  var data = require('jest-util');

  _jestUtil = function _jestUtil() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

var CustomConsole =
/*#__PURE__*/
function (_console$Console) {
  _inherits(CustomConsole, _console$Console);

  function CustomConsole(stdout, stderr) {
    var _this;

    var formatBuffer = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function (_type, message) {
      return message;
    };

    _classCallCheck(this, CustomConsole);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CustomConsole).call(this, stdout, stderr));

    _defineProperty(_assertThisInitialized(_this), '_stdout', void 0);

    _defineProperty(_assertThisInitialized(_this), '_stderr', void 0);

    _defineProperty(_assertThisInitialized(_this), '_formatBuffer', void 0);

    _defineProperty(_assertThisInitialized(_this), '_counters', {});

    _defineProperty(_assertThisInitialized(_this), '_timers', {});

    _defineProperty(_assertThisInitialized(_this), '_groupDepth', 0);

    _defineProperty(_assertThisInitialized(_this), 'Console', _console().Console);

    _this._stdout = stdout;
    _this._stderr = stderr;
    _this._formatBuffer = formatBuffer;
    return _this;
  }

  _createClass(CustomConsole, [{
    key: "_log",
    value: function _log(type, message) {
      (0, _jestUtil().clearLine)(this._stdout);

      _get(_getPrototypeOf(CustomConsole.prototype), "log", this).call(this, this._formatBuffer(type, '  '.repeat(this._groupDepth) + message));
    }
  }, {
    key: "_logError",
    value: function _logError(type, message) {
      (0, _jestUtil().clearLine)(this._stderr);

      _get(_getPrototypeOf(CustomConsole.prototype), "error", this).call(this, this._formatBuffer(type, '  '.repeat(this._groupDepth) + message));
    }
  }, {
    key: "assert",
    value: function assert(value, message) {
      try {
        (0, _assert()["default"])(value, message);
      } catch (error) {
        this._logError('assert', error.toString());
      }
    }
  }, {
    key: "count",
    value: function count() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      if (!this._counters[label]) {
        this._counters[label] = 0;
      }

      this._log('count', (0, _util().format)("".concat(label, ": ").concat(++this._counters[label])));
    }
  }, {
    key: "countReset",
    value: function countReset() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      this._counters[label] = 0;
    }
  }, {
    key: "debug",
    value: function debug(firstArg) {
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      this._log('debug', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "dir",
    value: function dir(firstArg) {
      for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }

      this._log('dir', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "dirxml",
    value: function dirxml(firstArg) {
      for (var _len3 = arguments.length, args = new Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
        args[_key3 - 1] = arguments[_key3];
      }

      this._log('dirxml', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "error",
    value: function error(firstArg) {
      for (var _len4 = arguments.length, args = new Array(_len4 > 1 ? _len4 - 1 : 0), _key4 = 1; _key4 < _len4; _key4++) {
        args[_key4 - 1] = arguments[_key4];
      }

      this._logError('error', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "group",
    value: function group(title) {
      this._groupDepth++;

      for (var _len5 = arguments.length, args = new Array(_len5 > 1 ? _len5 - 1 : 0), _key5 = 1; _key5 < _len5; _key5++) {
        args[_key5 - 1] = arguments[_key5];
      }

      if (title || args.length > 0) {
        this._log('group', _chalk()["default"].bold((0, _util().format).apply(void 0, [title].concat(args))));
      }
    }
  }, {
    key: "groupCollapsed",
    value: function groupCollapsed(title) {
      this._groupDepth++;

      for (var _len6 = arguments.length, args = new Array(_len6 > 1 ? _len6 - 1 : 0), _key6 = 1; _key6 < _len6; _key6++) {
        args[_key6 - 1] = arguments[_key6];
      }

      if (title || args.length > 0) {
        this._log('groupCollapsed', _chalk()["default"].bold((0, _util().format).apply(void 0, [title].concat(args))));
      }
    }
  }, {
    key: "groupEnd",
    value: function groupEnd() {
      if (this._groupDepth > 0) {
        this._groupDepth--;
      }
    }
  }, {
    key: "info",
    value: function info(firstArg) {
      for (var _len7 = arguments.length, args = new Array(_len7 > 1 ? _len7 - 1 : 0), _key7 = 1; _key7 < _len7; _key7++) {
        args[_key7 - 1] = arguments[_key7];
      }

      this._log('info', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "log",
    value: function log(firstArg) {
      for (var _len8 = arguments.length, args = new Array(_len8 > 1 ? _len8 - 1 : 0), _key8 = 1; _key8 < _len8; _key8++) {
        args[_key8 - 1] = arguments[_key8];
      }

      this._log('log', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "time",
    value: function time() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      if (this._timers[label]) {
        return;
      }

      this._timers[label] = new Date();
    }
  }, {
    key: "timeEnd",
    value: function timeEnd() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      var startTime = this._timers[label];

      if (startTime) {
        var endTime = new Date().getTime();
        var time = endTime - startTime.getTime();

        this._log('time', (0, _util().format)("".concat(label, ": ").concat((0, _jestUtil().formatTime)(time))));

        delete this._timers[label];
      }
    }
  }, {
    key: "timeLog",
    value: function timeLog() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      var startTime = this._timers[label];

      if (startTime) {
        var endTime = new Date();
        var time = endTime.getTime() - startTime.getTime();

        for (var _len9 = arguments.length, data = new Array(_len9 > 1 ? _len9 - 1 : 0), _key9 = 1; _key9 < _len9; _key9++) {
          data[_key9 - 1] = arguments[_key9];
        }

        this._log('time', (0, _util().format).apply(void 0, ["".concat(label, ": ").concat((0, _jestUtil().formatTime)(time))].concat(data)));
      }
    }
  }, {
    key: "warn",
    value: function warn(firstArg) {
      for (var _len10 = arguments.length, args = new Array(_len10 > 1 ? _len10 - 1 : 0), _key10 = 1; _key10 < _len10; _key10++) {
        args[_key10 - 1] = arguments[_key10];
      }

      this._logError('warn', (0, _util().format).apply(void 0, [firstArg].concat(args)));
    }
  }, {
    key: "getBuffer",
    value: function getBuffer() {
      return undefined;
    }
  }]);

  return CustomConsole;
}(_console().Console);

exports["default"] = CustomConsole;
