// jsdom/lib/jsdom/living/generated/dist/DOMStringMap.dev.js
"use strict";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;

var ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;
var interfaceName = "DOMStringMap";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'DOMStringMap'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["DOMStringMap"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor DOMStringMap is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

function makeProxy(wrapper, globalObject) {
  var proxyHandler = proxyHandlerCache.get(globalObject);

  if (proxyHandler === undefined) {
    proxyHandler = new ProxyHandler(globalObject);
    proxyHandlerCache.set(globalObject, proxyHandler);
  }

  return new Proxy(wrapper, proxyHandler);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper = makeProxy(wrapper, globalObject);
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper = makeProxy(wrapper, globalObject);
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  var DOMStringMap = function DOMStringMap() {
    _classCallCheck(this, DOMStringMap);

    throw new TypeError("Illegal constructor");
  };

  Object.defineProperties(DOMStringMap.prototype, _defineProperty({}, Symbol.toStringTag, {
    value: "DOMStringMap",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = DOMStringMap;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: DOMStringMap
  });
};

var proxyHandlerCache = new WeakMap();

var ProxyHandler =
/*#__PURE__*/
function () {
  function ProxyHandler(globalObject) {
    _classCallCheck(this, ProxyHandler);

    this._globalObject = globalObject;
  }

  _createClass(ProxyHandler, [{
    key: "get",
    value: function get(target, P, receiver) {
      if (_typeof(P) === "symbol") {
        return Reflect.get(target, P, receiver);
      }

      var desc = this.getOwnPropertyDescriptor(target, P);

      if (desc === undefined) {
        var parent = Object.getPrototypeOf(target);

        if (parent === null) {
          return undefined;
        }

        return Reflect.get(target, P, receiver);
      }

      if (!desc.get && !desc.set) {
        return desc.value;
      }

      var getter = desc.get;

      if (getter === undefined) {
        return undefined;
      }

      return Reflect.apply(getter, receiver, []);
    }
  }, {
    key: "has",
    value: function has(target, P) {
      if (_typeof(P) === "symbol") {
        return Reflect.has(target, P);
      }

      var desc = this.getOwnPropertyDescriptor(target, P);

      if (desc !== undefined) {
        return true;
      }

      var parent = Object.getPrototypeOf(target);

      if (parent !== null) {
        return Reflect.has(parent, P);
      }

      return false;
    }
  }, {
    key: "ownKeys",
    value: function ownKeys(target) {
      var keys = new Set();
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = target[implSymbol][utils.supportedPropertyNames][Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var key = _step.value;

          if (!utils.hasOwn(target, key)) {
            keys.add("".concat(key));
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = Reflect.ownKeys(target)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var _key = _step2.value;
          keys.add(_key);
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }

      return _toConsumableArray(keys);
    }
  }, {
    key: "getOwnPropertyDescriptor",
    value: function getOwnPropertyDescriptor(target, P) {
      if (_typeof(P) === "symbol") {
        return Reflect.getOwnPropertyDescriptor(target, P);
      }

      var ignoreNamedProps = false;
      var namedValue = target[implSymbol][utils.namedGet](P);

      if (namedValue !== undefined && !utils.hasOwn(target, P) && !ignoreNamedProps) {
        return {
          writable: true,
          enumerable: true,
          configurable: true,
          value: utils.tryWrapperForImpl(namedValue)
        };
      }

      return Reflect.getOwnPropertyDescriptor(target, P);
    }
  }, {
    key: "set",
    value: function set(target, P, V, receiver) {
      if (_typeof(P) === "symbol") {
        return Reflect.set(target, P, V, receiver);
      }

      if (target === receiver) {
        var globalObject = this._globalObject;

        if (typeof P === "string" && !utils.isArrayIndexPropName(P)) {
          var namedValue = V;
          namedValue = conversions["DOMString"](namedValue, {
            context: "Failed to set the '" + P + "' property on 'DOMStringMap': The provided value"
          });
          ceReactionsPreSteps_helpers_custom_elements(globalObject);

          try {
            var creating = !(target[implSymbol][utils.namedGet](P) !== undefined);

            if (creating) {
              target[implSymbol][utils.namedSetNew](P, namedValue);
            } else {
              target[implSymbol][utils.namedSetExisting](P, namedValue);
            }
          } finally {
            ceReactionsPostSteps_helpers_custom_elements(globalObject);
          }

          return true;
        }
      }

      var ownDesc;

      if (ownDesc === undefined) {
        ownDesc = Reflect.getOwnPropertyDescriptor(target, P);
      }

      if (ownDesc === undefined) {
        var parent = Reflect.getPrototypeOf(target);

        if (parent !== null) {
          return Reflect.set(parent, P, V, receiver);
        }

        ownDesc = {
          writable: true,
          enumerable: true,
          configurable: true,
          value: undefined
        };
      }

      if (!ownDesc.writable) {
        return false;
      }

      if (!utils.isObject(receiver)) {
        return false;
      }

      var existingDesc = Reflect.getOwnPropertyDescriptor(receiver, P);
      var valueDesc;

      if (existingDesc !== undefined) {
        if (existingDesc.get || existingDesc.set) {
          return false;
        }

        if (!existingDesc.writable) {
          return false;
        }

        valueDesc = {
          value: V
        };
      } else {
        valueDesc = {
          writable: true,
          enumerable: true,
          configurable: true,
          value: V
        };
      }

      return Reflect.defineProperty(receiver, P, valueDesc);
    }
  }, {
    key: "defineProperty",
    value: function defineProperty(target, P, desc) {
      if (_typeof(P) === "symbol") {
        return Reflect.defineProperty(target, P, desc);
      }

      var globalObject = this._globalObject;

      if (desc.get || desc.set) {
        return false;
      }

      var namedValue = desc.value;
      namedValue = conversions["DOMString"](namedValue, {
        context: "Failed to set the '" + P + "' property on 'DOMStringMap': The provided value"
      });
      ceReactionsPreSteps_helpers_custom_elements(globalObject);

      try {
        var creating = !(target[implSymbol][utils.namedGet](P) !== undefined);

        if (creating) {
          target[implSymbol][utils.namedSetNew](P, namedValue);
        } else {
          target[implSymbol][utils.namedSetExisting](P, namedValue);
        }
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }

      return true;
    }
  }, {
    key: "deleteProperty",
    value: function deleteProperty(target, P) {
      if (_typeof(P) === "symbol") {
        return Reflect.deleteProperty(target, P);
      }

      var globalObject = this._globalObject;

      if (target[implSymbol][utils.namedGet](P) !== undefined && !utils.hasOwn(target, P)) {
        ceReactionsPreSteps_helpers_custom_elements(globalObject);

        try {
          target[implSymbol][utils.namedDelete](P);
          return true;
        } finally {
          ceReactionsPostSteps_helpers_custom_elements(globalObject);
        }
      }

      return Reflect.deleteProperty(target, P);
    }
  }, {
    key: "preventExtensions",
    value: function preventExtensions() {
      return false;
    }
  }]);

  return ProxyHandler;
}();

var Impl = require("../nodes/DOMStringMap-impl.js");
