// jsdom/lib/jsdom/living/generated/dist/EventTarget.dev.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var EventListener = require("./EventListener.js");

var AddEventListenerOptions = require("./AddEventListenerOptions.js");

var EventListenerOptions = require("./EventListenerOptions.js");

var Event = require("./Event.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;
var interfaceName = "EventTarget";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'EventTarget'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["EventTarget"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor EventTarget is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window", "Worker", "AudioWorklet"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  var EventTarget =
  /*#__PURE__*/
  function () {
    function EventTarget() {
      _classCallCheck(this, EventTarget);

      return exports.setup(Object.create((this instanceof EventTarget ? this.constructor : void 0).prototype), globalObject, undefined);
    }

    _createClass(EventTarget, [{
      key: "addEventListener",
      value: function addEventListener(type, callback) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 2) {
          throw new TypeError("Failed to execute 'addEventListener' on 'EventTarget': 2 arguments required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg = arguments[1];

          if (_curArg === null || _curArg === undefined) {
            _curArg = null;
          } else {
            _curArg = EventListener.convert(_curArg, {
              context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 2"
            });
          }

          args.push(_curArg);
        }
        {
          var _curArg2 = arguments[2];

          if (_curArg2 !== undefined) {
            if (_curArg2 === null || _curArg2 === undefined) {
              _curArg2 = AddEventListenerOptions.convert(_curArg2, {
                context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 3"
              });
            } else if (utils.isObject(_curArg2)) {
              _curArg2 = AddEventListenerOptions.convert(_curArg2, {
                context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 3" + " dictionary"
              });
            } else if (typeof _curArg2 === "boolean") {
              _curArg2 = conversions["boolean"](_curArg2, {
                context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 3"
              });
            } else {
              _curArg2 = conversions["boolean"](_curArg2, {
                context: "Failed to execute 'addEventListener' on 'EventTarget': parameter 3"
              });
            }
          }

          args.push(_curArg2);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).addEventListener.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "removeEventListener",
      value: function removeEventListener(type, callback) {
        var _esValue$implSymbol2;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 2) {
          throw new TypeError("Failed to execute 'removeEventListener' on 'EventTarget': 2 arguments required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg3 = arguments[1];

          if (_curArg3 === null || _curArg3 === undefined) {
            _curArg3 = null;
          } else {
            _curArg3 = EventListener.convert(_curArg3, {
              context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 2"
            });
          }

          args.push(_curArg3);
        }
        {
          var _curArg4 = arguments[2];

          if (_curArg4 !== undefined) {
            if (_curArg4 === null || _curArg4 === undefined) {
              _curArg4 = EventListenerOptions.convert(_curArg4, {
                context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 3"
              });
            } else if (utils.isObject(_curArg4)) {
              _curArg4 = EventListenerOptions.convert(_curArg4, {
                context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 3" + " dictionary"
              });
            } else if (typeof _curArg4 === "boolean") {
              _curArg4 = conversions["boolean"](_curArg4, {
                context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 3"
              });
            } else {
              _curArg4 = conversions["boolean"](_curArg4, {
                context: "Failed to execute 'removeEventListener' on 'EventTarget': parameter 3"
              });
            }
          }

          args.push(_curArg4);
        }
        return (_esValue$implSymbol2 = esValue[implSymbol]).removeEventListener.apply(_esValue$implSymbol2, args);
      }
    }, {
      key: "dispatchEvent",
      value: function dispatchEvent(event) {
        var _esValue$implSymbol3;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'dispatchEvent' on 'EventTarget': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = Event.convert(curArg, {
            context: "Failed to execute 'dispatchEvent' on 'EventTarget': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol3 = esValue[implSymbol]).dispatchEvent.apply(_esValue$implSymbol3, args);
      }
    }]);

    return EventTarget;
  }();

  Object.defineProperties(EventTarget.prototype, _defineProperty({
    addEventListener: {
      enumerable: true
    },
    removeEventListener: {
      enumerable: true
    },
    dispatchEvent: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "EventTarget",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = EventTarget;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: EventTarget
  });
};

var Impl = require("../events/EventTarget-impl.js");
