// jsdom/lib/jsdom/living/generated/dist/KeyboardEvent.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var KeyboardEventInit = require("./KeyboardEventInit.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;

var UIEvent = require("./UIEvent.js");

var interfaceName = "KeyboardEvent";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'KeyboardEvent'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["KeyboardEvent"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor KeyboardEvent is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {
  UIEvent._internalSetup(wrapper, globalObject);
};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window"]);

exports.install = function (globalObject, globalNames) {
  var _Object$definePropert;

  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  if (globalObject.UIEvent === undefined) {
    throw new Error("Internal error: attempting to evaluate KeyboardEvent before UIEvent");
  }

  var KeyboardEvent =
  /*#__PURE__*/
  function (_globalObject$UIEvent) {
    _inherits(KeyboardEvent, _globalObject$UIEvent);

    function KeyboardEvent(type) {
      var _this;

      _classCallCheck(this, KeyboardEvent);

      if (arguments.length < 1) {
        throw new TypeError("Failed to construct 'KeyboardEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to construct 'KeyboardEvent': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];
        _curArg = KeyboardEventInit.convert(_curArg, {
          context: "Failed to construct 'KeyboardEvent': parameter 2"
        });
        args.push(_curArg);
      }
      return _possibleConstructorReturn(_this, exports.setup(Object.create((this instanceof KeyboardEvent ? this.constructor : void 0).prototype), globalObject, args));
    }

    _createClass(KeyboardEvent, [{
      key: "getModifierState",
      value: function getModifierState(keyArg) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'getModifierState' on 'KeyboardEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'getModifierState' on 'KeyboardEvent': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).getModifierState.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "initKeyboardEvent",
      value: function initKeyboardEvent(typeArg) {
        var _esValue$implSymbol2;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg2 = arguments[1];

          if (_curArg2 !== undefined) {
            _curArg2 = conversions["boolean"](_curArg2, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 2"
            });
          } else {
            _curArg2 = false;
          }

          args.push(_curArg2);
        }
        {
          var _curArg3 = arguments[2];

          if (_curArg3 !== undefined) {
            _curArg3 = conversions["boolean"](_curArg3, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 3"
            });
          } else {
            _curArg3 = false;
          }

          args.push(_curArg3);
        }
        {
          var _curArg4 = arguments[3];

          if (_curArg4 !== undefined) {
            if (_curArg4 === null || _curArg4 === undefined) {
              _curArg4 = null;
            } else {
              _curArg4 = utils.tryImplForWrapper(_curArg4);
            }
          } else {
            _curArg4 = null;
          }

          args.push(_curArg4);
        }
        {
          var _curArg5 = arguments[4];

          if (_curArg5 !== undefined) {
            _curArg5 = conversions["DOMString"](_curArg5, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 5"
            });
          } else {
            _curArg5 = "";
          }

          args.push(_curArg5);
        }
        {
          var _curArg6 = arguments[5];

          if (_curArg6 !== undefined) {
            _curArg6 = conversions["unsigned long"](_curArg6, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 6"
            });
          } else {
            _curArg6 = 0;
          }

          args.push(_curArg6);
        }
        {
          var _curArg7 = arguments[6];

          if (_curArg7 !== undefined) {
            _curArg7 = conversions["boolean"](_curArg7, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 7"
            });
          } else {
            _curArg7 = false;
          }

          args.push(_curArg7);
        }
        {
          var _curArg8 = arguments[7];

          if (_curArg8 !== undefined) {
            _curArg8 = conversions["boolean"](_curArg8, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 8"
            });
          } else {
            _curArg8 = false;
          }

          args.push(_curArg8);
        }
        {
          var _curArg9 = arguments[8];

          if (_curArg9 !== undefined) {
            _curArg9 = conversions["boolean"](_curArg9, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 9"
            });
          } else {
            _curArg9 = false;
          }

          args.push(_curArg9);
        }
        {
          var _curArg10 = arguments[9];

          if (_curArg10 !== undefined) {
            _curArg10 = conversions["boolean"](_curArg10, {
              context: "Failed to execute 'initKeyboardEvent' on 'KeyboardEvent': parameter 10"
            });
          } else {
            _curArg10 = false;
          }

          args.push(_curArg10);
        }
        return (_esValue$implSymbol2 = esValue[implSymbol]).initKeyboardEvent.apply(_esValue$implSymbol2, args);
      }
    }, {
      key: "key",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["key"];
      }
    }, {
      key: "code",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["code"];
      }
    }, {
      key: "location",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["location"];
      }
    }, {
      key: "ctrlKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["ctrlKey"];
      }
    }, {
      key: "shiftKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["shiftKey"];
      }
    }, {
      key: "altKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["altKey"];
      }
    }, {
      key: "metaKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["metaKey"];
      }
    }, {
      key: "repeat",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["repeat"];
      }
    }, {
      key: "isComposing",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["isComposing"];
      }
    }, {
      key: "charCode",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["charCode"];
      }
    }, {
      key: "keyCode",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["keyCode"];
      }
    }]);

    return KeyboardEvent;
  }(globalObject.UIEvent);

  Object.defineProperties(KeyboardEvent.prototype, (_Object$definePropert = {
    getModifierState: {
      enumerable: true
    },
    initKeyboardEvent: {
      enumerable: true
    },
    key: {
      enumerable: true
    },
    code: {
      enumerable: true
    },
    location: {
      enumerable: true
    },
    ctrlKey: {
      enumerable: true
    },
    shiftKey: {
      enumerable: true
    },
    altKey: {
      enumerable: true
    },
    metaKey: {
      enumerable: true
    },
    repeat: {
      enumerable: true
    },
    isComposing: {
      enumerable: true
    },
    charCode: {
      enumerable: true
    },
    keyCode: {
      enumerable: true
    }
  }, _defineProperty(_Object$definePropert, Symbol.toStringTag, {
    value: "KeyboardEvent",
    configurable: true
  }), _defineProperty(_Object$definePropert, "DOM_KEY_LOCATION_STANDARD", {
    value: 0x00,
    enumerable: true
  }), _defineProperty(_Object$definePropert, "DOM_KEY_LOCATION_LEFT", {
    value: 0x01,
    enumerable: true
  }), _defineProperty(_Object$definePropert, "DOM_KEY_LOCATION_RIGHT", {
    value: 0x02,
    enumerable: true
  }), _defineProperty(_Object$definePropert, "DOM_KEY_LOCATION_NUMPAD", {
    value: 0x03,
    enumerable: true
  }), _Object$definePropert));
  Object.defineProperties(KeyboardEvent, {
    DOM_KEY_LOCATION_STANDARD: {
      value: 0x00,
      enumerable: true
    },
    DOM_KEY_LOCATION_LEFT: {
      value: 0x01,
      enumerable: true
    },
    DOM_KEY_LOCATION_RIGHT: {
      value: 0x02,
      enumerable: true
    },
    DOM_KEY_LOCATION_NUMPAD: {
      value: 0x03,
      enumerable: true
    }
  });

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = KeyboardEvent;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: KeyboardEvent
  });
};

var Impl = require("../events/KeyboardEvent-impl.js");
