// jsdom/lib/jsdom/living/generated/dist/MessageEvent.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var MessageEventInit = require("./MessageEventInit.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;

var Event = require("./Event.js");

var interfaceName = "MessageEvent";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'MessageEvent'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["MessageEvent"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor MessageEvent is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {
  Event._internalSetup(wrapper, globalObject);
};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window", "Worker", "AudioWorklet"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  if (globalObject.Event === undefined) {
    throw new Error("Internal error: attempting to evaluate MessageEvent before Event");
  }

  var MessageEvent =
  /*#__PURE__*/
  function (_globalObject$Event) {
    _inherits(MessageEvent, _globalObject$Event);

    function MessageEvent(type) {
      var _this;

      _classCallCheck(this, MessageEvent);

      if (arguments.length < 1) {
        throw new TypeError("Failed to construct 'MessageEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to construct 'MessageEvent': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];
        _curArg = MessageEventInit.convert(_curArg, {
          context: "Failed to construct 'MessageEvent': parameter 2"
        });
        args.push(_curArg);
      }
      return _possibleConstructorReturn(_this, exports.setup(Object.create((this instanceof MessageEvent ? this.constructor : void 0).prototype), globalObject, args));
    }

    _createClass(MessageEvent, [{
      key: "initMessageEvent",
      value: function initMessageEvent(type) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'initMessageEvent' on 'MessageEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg2 = arguments[1];

          if (_curArg2 !== undefined) {
            _curArg2 = conversions["boolean"](_curArg2, {
              context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 2"
            });
          } else {
            _curArg2 = false;
          }

          args.push(_curArg2);
        }
        {
          var _curArg3 = arguments[2];

          if (_curArg3 !== undefined) {
            _curArg3 = conversions["boolean"](_curArg3, {
              context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 3"
            });
          } else {
            _curArg3 = false;
          }

          args.push(_curArg3);
        }
        {
          var _curArg4 = arguments[3];

          if (_curArg4 !== undefined) {
            _curArg4 = conversions["any"](_curArg4, {
              context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 4"
            });
          } else {
            _curArg4 = null;
          }

          args.push(_curArg4);
        }
        {
          var _curArg5 = arguments[4];

          if (_curArg5 !== undefined) {
            _curArg5 = conversions["USVString"](_curArg5, {
              context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 5"
            });
          } else {
            _curArg5 = "";
          }

          args.push(_curArg5);
        }
        {
          var _curArg6 = arguments[5];

          if (_curArg6 !== undefined) {
            _curArg6 = conversions["DOMString"](_curArg6, {
              context: "Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 6"
            });
          } else {
            _curArg6 = "";
          }

          args.push(_curArg6);
        }
        {
          var _curArg7 = arguments[6];

          if (_curArg7 !== undefined) {
            if (_curArg7 === null || _curArg7 === undefined) {
              _curArg7 = null;
            } else {
              _curArg7 = utils.tryImplForWrapper(_curArg7);
            }
          } else {
            _curArg7 = null;
          }

          args.push(_curArg7);
        }
        {
          var _curArg8 = arguments[7];

          if (_curArg8 !== undefined) {
            if (!utils.isObject(_curArg8)) {
              throw new TypeError("Failed to execute 'initMessageEvent' on 'MessageEvent': parameter 8" + " is not an iterable object.");
            } else {
              var V = [];
              var tmp = _curArg8;
              var _iteratorNormalCompletion = true;
              var _didIteratorError = false;
              var _iteratorError = undefined;

              try {
                for (var _iterator = tmp[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                  var nextItem = _step.value;
                  nextItem = utils.tryImplForWrapper(nextItem);
                  V.push(nextItem);
                }
              } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
              } finally {
                try {
                  if (!_iteratorNormalCompletion && _iterator["return"] != null) {
                    _iterator["return"]();
                  }
                } finally {
                  if (_didIteratorError) {
                    throw _iteratorError;
                  }
                }
              }

              _curArg8 = V;
            }
          } else {
            _curArg8 = [];
          }

          args.push(_curArg8);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).initMessageEvent.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "data",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["data"];
      }
    }, {
      key: "origin",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["origin"];
      }
    }, {
      key: "lastEventId",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["lastEventId"];
      }
    }, {
      key: "source",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return utils.tryWrapperForImpl(esValue[implSymbol]["source"]);
      }
    }, {
      key: "ports",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return utils.tryWrapperForImpl(esValue[implSymbol]["ports"]);
      }
    }]);

    return MessageEvent;
  }(globalObject.Event);

  Object.defineProperties(MessageEvent.prototype, _defineProperty({
    initMessageEvent: {
      enumerable: true
    },
    data: {
      enumerable: true
    },
    origin: {
      enumerable: true
    },
    lastEventId: {
      enumerable: true
    },
    source: {
      enumerable: true
    },
    ports: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "MessageEvent",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = MessageEvent;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: MessageEvent
  });
};

var Impl = require("../events/MessageEvent-impl.js");
