// jsdom/lib/jsdom/living/generated/dist/MouseEvent.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var MouseEventInit = require("./MouseEventInit.js");

var EventTarget = require("./EventTarget.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;

var UIEvent = require("./UIEvent.js");

var interfaceName = "MouseEvent";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'MouseEvent'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["MouseEvent"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor MouseEvent is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {
  UIEvent._internalSetup(wrapper, globalObject);
};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  if (globalObject.UIEvent === undefined) {
    throw new Error("Internal error: attempting to evaluate MouseEvent before UIEvent");
  }

  var MouseEvent =
  /*#__PURE__*/
  function (_globalObject$UIEvent) {
    _inherits(MouseEvent, _globalObject$UIEvent);

    function MouseEvent(type) {
      var _this;

      _classCallCheck(this, MouseEvent);

      if (arguments.length < 1) {
        throw new TypeError("Failed to construct 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to construct 'MouseEvent': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];
        _curArg = MouseEventInit.convert(_curArg, {
          context: "Failed to construct 'MouseEvent': parameter 2"
        });
        args.push(_curArg);
      }
      return _possibleConstructorReturn(_this, exports.setup(Object.create((this instanceof MouseEvent ? this.constructor : void 0).prototype), globalObject, args));
    }

    _createClass(MouseEvent, [{
      key: "getModifierState",
      value: function getModifierState(keyArg) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'getModifierState' on 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'getModifierState' on 'MouseEvent': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).getModifierState.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "initMouseEvent",
      value: function initMouseEvent(typeArg) {
        var _esValue$implSymbol2;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'initMouseEvent' on 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg2 = arguments[1];

          if (_curArg2 !== undefined) {
            _curArg2 = conversions["boolean"](_curArg2, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 2"
            });
          } else {
            _curArg2 = false;
          }

          args.push(_curArg2);
        }
        {
          var _curArg3 = arguments[2];

          if (_curArg3 !== undefined) {
            _curArg3 = conversions["boolean"](_curArg3, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 3"
            });
          } else {
            _curArg3 = false;
          }

          args.push(_curArg3);
        }
        {
          var _curArg4 = arguments[3];

          if (_curArg4 !== undefined) {
            if (_curArg4 === null || _curArg4 === undefined) {
              _curArg4 = null;
            } else {
              _curArg4 = utils.tryImplForWrapper(_curArg4);
            }
          } else {
            _curArg4 = null;
          }

          args.push(_curArg4);
        }
        {
          var _curArg5 = arguments[4];

          if (_curArg5 !== undefined) {
            _curArg5 = conversions["long"](_curArg5, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 5"
            });
          } else {
            _curArg5 = 0;
          }

          args.push(_curArg5);
        }
        {
          var _curArg6 = arguments[5];

          if (_curArg6 !== undefined) {
            _curArg6 = conversions["long"](_curArg6, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 6"
            });
          } else {
            _curArg6 = 0;
          }

          args.push(_curArg6);
        }
        {
          var _curArg7 = arguments[6];

          if (_curArg7 !== undefined) {
            _curArg7 = conversions["long"](_curArg7, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 7"
            });
          } else {
            _curArg7 = 0;
          }

          args.push(_curArg7);
        }
        {
          var _curArg8 = arguments[7];

          if (_curArg8 !== undefined) {
            _curArg8 = conversions["long"](_curArg8, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 8"
            });
          } else {
            _curArg8 = 0;
          }

          args.push(_curArg8);
        }
        {
          var _curArg9 = arguments[8];

          if (_curArg9 !== undefined) {
            _curArg9 = conversions["long"](_curArg9, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 9"
            });
          } else {
            _curArg9 = 0;
          }

          args.push(_curArg9);
        }
        {
          var _curArg10 = arguments[9];

          if (_curArg10 !== undefined) {
            _curArg10 = conversions["boolean"](_curArg10, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 10"
            });
          } else {
            _curArg10 = 0;
          }

          args.push(_curArg10);
        }
        {
          var _curArg11 = arguments[10];

          if (_curArg11 !== undefined) {
            _curArg11 = conversions["boolean"](_curArg11, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 11"
            });
          } else {
            _curArg11 = 0;
          }

          args.push(_curArg11);
        }
        {
          var _curArg12 = arguments[11];

          if (_curArg12 !== undefined) {
            _curArg12 = conversions["boolean"](_curArg12, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 12"
            });
          } else {
            _curArg12 = 0;
          }

          args.push(_curArg12);
        }
        {
          var _curArg13 = arguments[12];

          if (_curArg13 !== undefined) {
            _curArg13 = conversions["boolean"](_curArg13, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 13"
            });
          } else {
            _curArg13 = 0;
          }

          args.push(_curArg13);
        }
        {
          var _curArg14 = arguments[13];

          if (_curArg14 !== undefined) {
            _curArg14 = conversions["short"](_curArg14, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 14"
            });
          } else {
            _curArg14 = 0;
          }

          args.push(_curArg14);
        }
        {
          var _curArg15 = arguments[14];

          if (_curArg15 !== undefined) {
            if (_curArg15 === null || _curArg15 === undefined) {
              _curArg15 = null;
            } else {
              _curArg15 = EventTarget.convert(_curArg15, {
                context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 15"
              });
            }
          } else {
            _curArg15 = null;
          }

          args.push(_curArg15);
        }
        return (_esValue$implSymbol2 = esValue[implSymbol]).initMouseEvent.apply(_esValue$implSymbol2, args);
      }
    }, {
      key: "screenX",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["screenX"];
      }
    }, {
      key: "screenY",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["screenY"];
      }
    }, {
      key: "clientX",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["clientX"];
      }
    }, {
      key: "clientY",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["clientY"];
      }
    }, {
      key: "ctrlKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["ctrlKey"];
      }
    }, {
      key: "shiftKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["shiftKey"];
      }
    }, {
      key: "altKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["altKey"];
      }
    }, {
      key: "metaKey",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["metaKey"];
      }
    }, {
      key: "button",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["button"];
      }
    }, {
      key: "buttons",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["buttons"];
      }
    }, {
      key: "relatedTarget",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return utils.tryWrapperForImpl(esValue[implSymbol]["relatedTarget"]);
      }
    }]);

    return MouseEvent;
  }(globalObject.UIEvent);

  Object.defineProperties(MouseEvent.prototype, _defineProperty({
    getModifierState: {
      enumerable: true
    },
    initMouseEvent: {
      enumerable: true
    },
    screenX: {
      enumerable: true
    },
    screenY: {
      enumerable: true
    },
    clientX: {
      enumerable: true
    },
    clientY: {
      enumerable: true
    },
    ctrlKey: {
      enumerable: true
    },
    shiftKey: {
      enumerable: true
    },
    altKey: {
      enumerable: true
    },
    metaKey: {
      enumerable: true
    },
    button: {
      enumerable: true
    },
    buttons: {
      enumerable: true
    },
    relatedTarget: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "MouseEvent",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = MouseEvent;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: MouseEvent
  });
};

var Impl = require("../events/MouseEvent-impl.js");
