// jsdom/lib/jsdom/living/generated/dist/Storage.dev.js
"use strict";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;
var interfaceName = "Storage";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'Storage'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["Storage"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor Storage is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper = new Proxy(wrapper, proxyHandler);
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper = new Proxy(wrapper, proxyHandler);
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  var Storage =
  /*#__PURE__*/
  function () {
    function Storage() {
      _classCallCheck(this, Storage);

      throw new TypeError("Illegal constructor");
    }

    _createClass(Storage, [{
      key: "key",
      value: function key(index) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'key' on 'Storage': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["unsigned long"](curArg, {
            context: "Failed to execute 'key' on 'Storage': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).key.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "getItem",
      value: function getItem(key) {
        var _esValue$implSymbol2;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'getItem' on 'Storage': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'getItem' on 'Storage': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol2 = esValue[implSymbol]).getItem.apply(_esValue$implSymbol2, args);
      }
    }, {
      key: "setItem",
      value: function setItem(key, value) {
        var _esValue$implSymbol3;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 2) {
          throw new TypeError("Failed to execute 'setItem' on 'Storage': 2 arguments required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'setItem' on 'Storage': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg = arguments[1];
          _curArg = conversions["DOMString"](_curArg, {
            context: "Failed to execute 'setItem' on 'Storage': parameter 2"
          });
          args.push(_curArg);
        }
        return (_esValue$implSymbol3 = esValue[implSymbol]).setItem.apply(_esValue$implSymbol3, args);
      }
    }, {
      key: "removeItem",
      value: function removeItem(key) {
        var _esValue$implSymbol4;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'removeItem' on 'Storage': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'removeItem' on 'Storage': parameter 1"
          });
          args.push(curArg);
        }
        return (_esValue$implSymbol4 = esValue[implSymbol]).removeItem.apply(_esValue$implSymbol4, args);
      }
    }, {
      key: "clear",
      value: function clear() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol].clear();
      }
    }, {
      key: "length",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["length"];
      }
    }]);

    return Storage;
  }();

  Object.defineProperties(Storage.prototype, _defineProperty({
    key: {
      enumerable: true
    },
    getItem: {
      enumerable: true
    },
    setItem: {
      enumerable: true
    },
    removeItem: {
      enumerable: true
    },
    clear: {
      enumerable: true
    },
    length: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "Storage",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = Storage;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: Storage
  });
};

var proxyHandler = {
  get: function get(target, P, receiver) {
    if (_typeof(P) === "symbol") {
      return Reflect.get(target, P, receiver);
    }

    var desc = this.getOwnPropertyDescriptor(target, P);

    if (desc === undefined) {
      var parent = Object.getPrototypeOf(target);

      if (parent === null) {
        return undefined;
      }

      return Reflect.get(target, P, receiver);
    }

    if (!desc.get && !desc.set) {
      return desc.value;
    }

    var getter = desc.get;

    if (getter === undefined) {
      return undefined;
    }

    return Reflect.apply(getter, receiver, []);
  },
  has: function has(target, P) {
    if (_typeof(P) === "symbol") {
      return Reflect.has(target, P);
    }

    var desc = this.getOwnPropertyDescriptor(target, P);

    if (desc !== undefined) {
      return true;
    }

    var parent = Object.getPrototypeOf(target);

    if (parent !== null) {
      return Reflect.has(parent, P);
    }

    return false;
  },
  ownKeys: function ownKeys(target) {
    var keys = new Set();
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = target[implSymbol][utils.supportedPropertyNames][Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var key = _step.value;

        if (!(key in target)) {
          keys.add("".concat(key));
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = Reflect.ownKeys(target)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var _key = _step2.value;
        keys.add(_key);
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
          _iterator2["return"]();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }

    return _toConsumableArray(keys);
  },
  getOwnPropertyDescriptor: function getOwnPropertyDescriptor(target, P) {
    if (_typeof(P) === "symbol") {
      return Reflect.getOwnPropertyDescriptor(target, P);
    }

    var ignoreNamedProps = false;
    var namedValue = target[implSymbol].getItem(P);

    if (namedValue !== null && !(P in target) && !ignoreNamedProps) {
      return {
        writable: true,
        enumerable: true,
        configurable: true,
        value: utils.tryWrapperForImpl(namedValue)
      };
    }

    return Reflect.getOwnPropertyDescriptor(target, P);
  },
  set: function set(target, P, V, receiver) {
    if (_typeof(P) === "symbol") {
      return Reflect.set(target, P, V, receiver);
    }

    if (target === receiver) {
      if (typeof P === "string" && !utils.isArrayIndexPropName(P)) {
        var namedValue = V;
        namedValue = conversions["DOMString"](namedValue, {
          context: "Failed to set the '" + P + "' property on 'Storage': The provided value"
        });
        target[implSymbol].setItem(P, namedValue);
        return true;
      }
    }

    var ownDesc;

    if (ownDesc === undefined) {
      ownDesc = Reflect.getOwnPropertyDescriptor(target, P);
    }

    if (ownDesc === undefined) {
      var parent = Reflect.getPrototypeOf(target);

      if (parent !== null) {
        return Reflect.set(parent, P, V, receiver);
      }

      ownDesc = {
        writable: true,
        enumerable: true,
        configurable: true,
        value: undefined
      };
    }

    if (!ownDesc.writable) {
      return false;
    }

    if (!utils.isObject(receiver)) {
      return false;
    }

    var existingDesc = Reflect.getOwnPropertyDescriptor(receiver, P);
    var valueDesc;

    if (existingDesc !== undefined) {
      if (existingDesc.get || existingDesc.set) {
        return false;
      }

      if (!existingDesc.writable) {
        return false;
      }

      valueDesc = {
        value: V
      };
    } else {
      valueDesc = {
        writable: true,
        enumerable: true,
        configurable: true,
        value: V
      };
    }

    return Reflect.defineProperty(receiver, P, valueDesc);
  },
  defineProperty: function defineProperty(target, P, desc) {
    if (_typeof(P) === "symbol") {
      return Reflect.defineProperty(target, P, desc);
    }

    if (!utils.hasOwn(target, P)) {
      if (desc.get || desc.set) {
        return false;
      }

      var namedValue = desc.value;
      namedValue = conversions["DOMString"](namedValue, {
        context: "Failed to set the '" + P + "' property on 'Storage': The provided value"
      });
      target[implSymbol].setItem(P, namedValue);
      return true;
    }

    return Reflect.defineProperty(target, P, desc);
  },
  deleteProperty: function deleteProperty(target, P) {
    if (_typeof(P) === "symbol") {
      return Reflect.deleteProperty(target, P);
    }

    if (target[implSymbol].getItem(P) !== null && !(P in target)) {
      target[implSymbol].removeItem(P);
      return true;
    }

    return Reflect.deleteProperty(target, P);
  },
  preventExtensions: function preventExtensions() {
    return false;
  }
};

var Impl = require("../webstorage/Storage-impl.js");
