// jsdom/lib/jsdom/living/generated/dist/UIEvent.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var UIEventInit = require("./UIEventInit.js");

var implSymbol = utils.implSymbol;
var ctorRegistrySymbol = utils.ctorRegistrySymbol;

var Event = require("./Event.js");

var interfaceName = "UIEvent";

exports.is = function (value) {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};

exports.isImpl = function (value) {
  return utils.isObject(value) && value instanceof Impl.implementation;
};

exports.convert = function (value) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$context = _ref.context,
      context = _ref$context === void 0 ? "The provided value" : _ref$context;

  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }

  throw new TypeError("".concat(context, " is not of type 'UIEvent'."));
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  var ctor = globalObject[ctorRegistrySymbol]["UIEvent"];

  if (ctor === undefined) {
    throw new Error("Internal error: constructor UIEvent is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = function (globalObject, constructorArgs, privateData) {
  var wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = function (globalObject, constructorArgs, privateData) {
  var wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = function (wrapper, globalObject) {
  Event._internalSetup(wrapper, globalObject);
};

exports.setup = function (wrapper, globalObject) {
  var constructorArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var privateData = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper;
};

exports["new"] = function (globalObject) {
  var wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);

  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });
  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;

  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }

  return wrapper[implSymbol];
};

var exposed = new Set(["Window"]);

exports.install = function (globalObject, globalNames) {
  if (!globalNames.some(function (globalName) {
    return exposed.has(globalName);
  })) {
    return;
  }

  if (globalObject.Event === undefined) {
    throw new Error("Internal error: attempting to evaluate UIEvent before Event");
  }

  var UIEvent =
  /*#__PURE__*/
  function (_globalObject$Event) {
    _inherits(UIEvent, _globalObject$Event);

    function UIEvent(type) {
      var _this;

      _classCallCheck(this, UIEvent);

      if (arguments.length < 1) {
        throw new TypeError("Failed to construct 'UIEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to construct 'UIEvent': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];
        _curArg = UIEventInit.convert(_curArg, {
          context: "Failed to construct 'UIEvent': parameter 2"
        });
        args.push(_curArg);
      }
      return _possibleConstructorReturn(_this, exports.setup(Object.create((this instanceof UIEvent ? this.constructor : void 0).prototype), globalObject, args));
    }

    _createClass(UIEvent, [{
      key: "initUIEvent",
      value: function initUIEvent(typeArg) {
        var _esValue$implSymbol;

        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        if (arguments.length < 1) {
          throw new TypeError("Failed to execute 'initUIEvent' on 'UIEvent': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'initUIEvent' on 'UIEvent': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg2 = arguments[1];

          if (_curArg2 !== undefined) {
            _curArg2 = conversions["boolean"](_curArg2, {
              context: "Failed to execute 'initUIEvent' on 'UIEvent': parameter 2"
            });
          } else {
            _curArg2 = false;
          }

          args.push(_curArg2);
        }
        {
          var _curArg3 = arguments[2];

          if (_curArg3 !== undefined) {
            _curArg3 = conversions["boolean"](_curArg3, {
              context: "Failed to execute 'initUIEvent' on 'UIEvent': parameter 3"
            });
          } else {
            _curArg3 = false;
          }

          args.push(_curArg3);
        }
        {
          var _curArg4 = arguments[3];

          if (_curArg4 !== undefined) {
            if (_curArg4 === null || _curArg4 === undefined) {
              _curArg4 = null;
            } else {
              _curArg4 = utils.tryImplForWrapper(_curArg4);
            }
          } else {
            _curArg4 = null;
          }

          args.push(_curArg4);
        }
        {
          var _curArg5 = arguments[4];

          if (_curArg5 !== undefined) {
            _curArg5 = conversions["long"](_curArg5, {
              context: "Failed to execute 'initUIEvent' on 'UIEvent': parameter 5"
            });
          } else {
            _curArg5 = 0;
          }

          args.push(_curArg5);
        }
        return (_esValue$implSymbol = esValue[implSymbol]).initUIEvent.apply(_esValue$implSymbol, args);
      }
    }, {
      key: "view",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return utils.tryWrapperForImpl(esValue[implSymbol]["view"]);
      }
    }, {
      key: "detail",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["detail"];
      }
    }, {
      key: "which",
      get: function get() {
        var esValue = this !== null && this !== undefined ? this : globalObject;

        if (!exports.is(esValue)) {
          throw new TypeError("Illegal invocation");
        }

        return esValue[implSymbol]["which"];
      }
    }]);

    return UIEvent;
  }(globalObject.Event);

  Object.defineProperties(UIEvent.prototype, _defineProperty({
    initUIEvent: {
      enumerable: true
    },
    view: {
      enumerable: true
    },
    detail: {
      enumerable: true
    },
    which: {
      enumerable: true
    }
  }, Symbol.toStringTag, {
    value: "UIEvent",
    configurable: true
  }));

  if (globalObject[ctorRegistrySymbol] === undefined) {
    globalObject[ctorRegistrySymbol] = Object.create(null);
  }

  globalObject[ctorRegistrySymbol][interfaceName] = UIEvent;
  Object.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: UIEvent
  });
};

var Impl = require("../events/UIEvent-impl.js");
