// jsdom/lib/jsdom/living/helpers/dist/document-base-url.dev.js
"use strict";

var whatwgURL = require("whatwg-url");

exports.documentBaseURL = function (document) {
  // https://html.spec.whatwg.org/multipage/infrastructure.html#document-base-url
  var firstBase = document.querySelector("base[href]");
  var fallbackBaseURL = exports.fallbackBaseURL(document);

  if (firstBase === null) {
    return fallbackBaseURL;
  }

  return frozenBaseURL(firstBase, fallbackBaseURL);
};

exports.documentBaseURLSerialized = function (document) {
  return whatwgURL.serializeURL(exports.documentBaseURL(document));
};

exports.fallbackBaseURL = function (document) {
  // https://html.spec.whatwg.org/multipage/infrastructure.html#fallback-base-url
  // Unimplemented: <iframe srcdoc>
  if (document.URL === "about:blank" && document._defaultView && document._defaultView._parent !== document._defaultView) {
    return exports.documentBaseURL(document._defaultView._parent._document);
  }

  return document._URL;
};

exports.parseURLToResultingURLRecord = function (url, document) {
  // https://html.spec.whatwg.org/#resolve-a-url
  // Encoding stuff ignored; always UTF-8 for us, for now.
  var baseURL = exports.documentBaseURL(document);
  return whatwgURL.parseURL(url, {
    baseURL: baseURL
  }); // This returns the resulting URL record; to get the resulting URL string, just serialize it.
};

function frozenBaseURL(baseElement, fallbackBaseURL) {
  // https://html.spec.whatwg.org/multipage/semantics.html#frozen-base-url
  // The spec is eager (setting the frozen base URL when things change); we are lazy (getting it when we need to)
  var baseHrefAttribute = baseElement.getAttributeNS(null, "href");
  var result = whatwgURL.parseURL(baseHrefAttribute, {
    baseURL: fallbackBaseURL
  });
  return result === null ? fallbackBaseURL : result;
}
