// jsdom/lib/jsdom/living/helpers/dist/style-rules.dev.js
"use strict";

var cssom = require("cssom");

var defaultStyleSheet = require("../../browser/default-stylesheet");

var _require = require("./selectors"),
    matchesDontThrow = _require.matchesDontThrow;

var _Array$prototype = Array.prototype,
    forEach = _Array$prototype.forEach,
    indexOf = _Array$prototype.indexOf;
var parsedDefaultStyleSheet; // Properties for which getResolvedValue is implemented. This is less than
// every supported property.
// https://drafts.csswg.org/indexes/#properties

exports.propertiesWithResolvedValueImplemented = {
  __proto__: null,
  // https://drafts.csswg.org/css2/visufx.html#visibility
  visibility: {
    inherited: true,
    initial: "visible",
    computedValue: "as-specified"
  }
};

exports.forEachMatchingSheetRuleOfElement = function (elementImpl, handleRule) {
  function handleSheet(sheet) {
    forEach.call(sheet.cssRules, function (rule) {
      if (rule.media) {
        if (indexOf.call(rule.media, "screen") !== -1) {
          forEach.call(rule.cssRules, function (innerRule) {
            if (matches(innerRule, elementImpl)) {
              handleRule(innerRule);
            }
          });
        }
      } else if (matches(rule, elementImpl)) {
        handleRule(rule);
      }
    });
  }

  if (!parsedDefaultStyleSheet) {
    parsedDefaultStyleSheet = cssom.parse(defaultStyleSheet);
  }

  handleSheet(parsedDefaultStyleSheet);
  forEach.call(elementImpl._ownerDocument.styleSheets._list, handleSheet);
};

function matches(rule, element) {
  return matchesDontThrow(element, rule.selectorText);
} // Naive implementation of https://drafts.csswg.org/css-cascade-4/#cascading
// based on the previous jsdom implementation of getComputedStyle.
// Does not implement https://drafts.csswg.org/css-cascade-4/#cascade-specificity,
// or rather specificity is only implemented by the order in which the matching
// rules appear. The last rule is the most specific while the first rule is
// the least specific.


function getCascadedPropertyValue(element, property) {
  var value = "";
  exports.forEachMatchingSheetRuleOfElement(element, function (rule) {
    value = rule.style.getPropertyValue(property);
  });
  var inlineValue = element.style.getPropertyValue(property);

  if (inlineValue !== "" && inlineValue !== null) {
    value = inlineValue;
  }

  return value;
} // https://drafts.csswg.org/css-cascade-4/#specified-value


function getSpecifiedValue(element, property) {
  var cascade = getCascadedPropertyValue(element, property);

  if (cascade !== "") {
    return cascade;
  } // Defaulting


  var _exports$propertiesWi = exports.propertiesWithResolvedValueImplemented[property],
      initial = _exports$propertiesWi.initial,
      inherited = _exports$propertiesWi.inherited;

  if (inherited && element.parentElement !== null) {
    return getComputedValue(element.parentElement, property);
  } // root element without parent element or inherited property


  return initial;
} // https://drafts.csswg.org/css-cascade-4/#computed-value


function getComputedValue(element, property) {
  var computedValue = exports.propertiesWithResolvedValueImplemented[property].computedValue;

  if (computedValue === "as-specified") {
    return getSpecifiedValue(element, property);
  }

  throw new TypeError("Internal error: unrecognized computed value instruction '".concat(computedValue, "'"));
} // https://drafts.csswg.org/cssom/#resolved-value
// Only implements `visibility`


exports.getResolvedValue = function (element, property) {
  // Determined for special case properties, none of which are implemented here.
  // So we skip to "any other property: The resolved value is the computed value."
  return getComputedValue(element, property);
};

exports.SHADOW_DOM_PSEUDO_REGEXP = /^::(?:part|slotted)\(/i;
