// rsvp/lib/rsvp/dist/instrument.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = instrument;

var _config = require("./config");

var queue = [];

function scheduleFlush() {
  setTimeout(function () {
    for (var i = 0; i < queue.length; i++) {
      var entry = queue[i];
      var payload = entry.payload;
      payload.guid = payload.key + payload.id;
      payload.childGuid = payload.key + payload.childId;

      if (payload.error) {
        payload.stack = payload.error.stack;
      }

      _config.config['trigger'](entry.name, entry.payload);
    }

    queue.length = 0;
  }, 50);
}

function instrument(eventName, promise, child) {
  if (1 === queue.push({
    name: eventName,
    payload: {
      key: promise._guidKey,
      id: promise._id,
      eventName: eventName,
      detail: promise._result,
      childId: child && child._id,
      label: promise._label,
      timeStamp: Date.now(),
      error: _config.config["instrument-with-stack"] ? new Error(promise._label) : null
    }
  })) {
    scheduleFlush();
  }
}
