// jest-snapshot/build/dist/colors.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.bBackground3 = exports.bForeground3 = exports.aBackground3 = exports.aForeground3 = exports.bBackground2 = exports.bForeground2 = exports.aBackground2 = exports.aForeground2 = void 0;
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// https://jonasjacek.github.io/colors/

var aForeground2 = 90;
exports.aForeground2 = aForeground2;
var aBackground2 = 225;
exports.aBackground2 = aBackground2;
var bForeground2 = 23;
exports.bForeground2 = bForeground2;
var bBackground2 = 195;
exports.bBackground2 = bBackground2;
var aForeground3 = [0x80, 0, 0x80];
exports.aForeground3 = aForeground3;
var aBackground3 = [0xff, 0xd7, 0xff];
exports.aBackground3 = aBackground3;
var bForeground3 = [0, 0x5f, 0x5f];
exports.bForeground3 = bForeground3;
var bBackground3 = [0xd7, 0xff, 0xff];
exports.bBackground3 = bBackground3;
