// sisteransi/src/dist/index.dev.js
'use strict';

var ESC = '\x1B';
var CSI = "".concat(ESC, "[");
var beep = "\x07";
var cursor = {
  to: function to(x, y) {
    if (!y) return "".concat(CSI).concat(x + 1, "G");
    return "".concat(CSI).concat(y + 1, ";").concat(x + 1, "H");
  },
  move: function move(x, y) {
    var ret = '';
    if (x < 0) ret += "".concat(CSI).concat(-x, "D");else if (x > 0) ret += "".concat(CSI).concat(x, "C");
    if (y < 0) ret += "".concat(CSI).concat(-y, "A");else if (y > 0) ret += "".concat(CSI).concat(y, "B");
    return ret;
  },
  up: function up() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI).concat(count, "A");
  },
  down: function down() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI).concat(count, "B");
  },
  forward: function forward() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI).concat(count, "C");
  },
  backward: function backward() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI).concat(count, "D");
  },
  nextLine: function nextLine() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "E").repeat(count);
  },
  prevLine: function prevLine() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "F").repeat(count);
  },
  left: "".concat(CSI, "G"),
  hide: "".concat(CSI, "?25l"),
  show: "".concat(CSI, "?25h"),
  save: "".concat(ESC, "7"),
  restore: "".concat(ESC, "8")
};
var scroll = {
  up: function up() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "S").repeat(count);
  },
  down: function down() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "T").repeat(count);
  }
};
var erase = {
  screen: "".concat(CSI, "2J"),
  up: function up() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "1J").repeat(count);
  },
  down: function down() {
    var count = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
    return "".concat(CSI, "J").repeat(count);
  },
  line: "".concat(CSI, "2K"),
  lineEnd: "".concat(CSI, "K"),
  lineStart: "".concat(CSI, "1K"),
  lines: function lines(count) {
    var clear = '';

    for (var i = 0; i < count; i++) {
      clear += this.line + (i < count - 1 ? cursor.up() : '');
    }

    if (count) clear += cursor.left;
    return clear;
  }
};
module.exports = {
  cursor: cursor,
  scroll: scroll,
  erase: erase,
  beep: beep
};
