// jest-resolve/build/dist/shouldLoadAsEsm.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.clearCachedLookups = clearCachedLookups;
exports["default"] = cachedShouldLoadAsEsm;

function _path() {
  var data = require('path');

  _path = function _path() {
    return data;
  };

  return data;
}

function _vm() {
  var data = require('vm');

  _vm = function _vm() {
    return data;
  };

  return data;
}

function _readPkgUp() {
  var data = _interopRequireDefault(require('read-pkg-up'));

  _readPkgUp = function _readPkgUp() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @ts-expect-error: experimental, not added to the types


var runtimeSupportsVmModules = typeof _vm().SyntheticModule === 'function';
var cachedFileLookups = new Map();
var cachedDirLookups = new Map();

function clearCachedLookups() {
  cachedFileLookups.clear();
  cachedDirLookups.clear();
}

function cachedShouldLoadAsEsm(path) {
  var cachedLookup = cachedFileLookups.get(path);

  if (cachedLookup === undefined) {
    cachedLookup = shouldLoadAsEsm(path);
    cachedFileLookups.set(path, cachedLookup);
  }

  return cachedLookup;
} // this is a bad version of what https://github.com/nodejs/modules/issues/393 would provide


function shouldLoadAsEsm(path) {
  if (!runtimeSupportsVmModules) {
    return false;
  }

  var extension = (0, _path().extname)(path);

  if (extension === '.mjs') {
    return true;
  }

  if (extension === '.cjs') {
    return false;
  } // this isn't correct - we might wanna load any file as a module (using synthetic module)
  // do we need an option to Jest so people can opt in to ESM for non-js?


  if (extension !== '.js') {
    return false;
  }

  var cwd = (0, _path().dirname)(path);
  var cachedLookup = cachedDirLookups.get(cwd);

  if (cachedLookup === undefined) {
    cachedLookup = cachedPkgCheck(cwd);
    cachedFileLookups.set(cwd, cachedLookup);
  }

  return cachedLookup;
}

function cachedPkgCheck(cwd) {
  // TODO: can we cache lookups somehow?
  var pkg = _readPkgUp()["default"].sync({
    cwd: cwd,
    normalize: false
  });

  if (!pkg) {
    return false;
  }

  return pkg.packageJson.type === 'module';
}
