// resolve-from/dist/index.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var path = require('path');

var Module = require('module');

var fs = require('fs');

var resolveFrom = function resolveFrom(fromDirectory, moduleId, silent) {
  if (typeof fromDirectory !== 'string') {
    throw new TypeError("Expected `fromDir` to be of type `string`, got `".concat(_typeof(fromDirectory), "`"));
  }

  if (typeof moduleId !== 'string') {
    throw new TypeError("Expected `moduleId` to be of type `string`, got `".concat(_typeof(moduleId), "`"));
  }

  try {
    fromDirectory = fs.realpathSync(fromDirectory);
  } catch (error) {
    if (error.code === 'ENOENT') {
      fromDirectory = path.resolve(fromDirectory);
    } else if (silent) {
      return;
    } else {
      throw error;
    }
  }

  var fromFile = path.join(fromDirectory, 'noop.js');

  var resolveFileName = function resolveFileName() {
    return Module._resolveFilename(moduleId, {
      id: fromFile,
      filename: fromFile,
      paths: Module._nodeModulePaths(fromDirectory)
    });
  };

  if (silent) {
    try {
      return resolveFileName();
    } catch (error) {
      return;
    }
  }

  return resolveFileName();
};

module.exports = function (fromDirectory, moduleId) {
  return resolveFrom(fromDirectory, moduleId);
};

module.exports.silent = function (fromDirectory, moduleId) {
  return resolveFrom(fromDirectory, moduleId, true);
};
